datreant/backends/statefiles.py

```diff
@@ -0,0 +1,448 @@
+                values must be single numbers, strings, or boolean values;
+                values that are not these types are not added
+
+        """
+        for key, value in categories.items():
+            if (isinstance(value, (int, float, string_types, bool)) or
+                    value is None):
+                self._record['categories'][key] = value
+
+    @FileSerial._write
+    @FileSerial._pull_push
+    def del_categories(self, *categories, **kwargs):
+        """Delete categories from Treant.
+
+        Any number of categories (keys) can be given as arguments, and these
+        keys (with their values) will be deleted.
+
+        :Arguments:
+            *categories*
+                Categories to delete.
+
+        :Keywords:
+            *all*
+                When True, delete all categories [``False``]
+
+        """
+        purge = kwargs.pop('all', False)
+
+        if purge:
+            self._record['categories'] = dict()
+        else:
+            for key in categories:
+                # continue even if key not already present
+                self._record['categories'].pop(key, None)
+
+
+class GroupFile(TreantFile):
+    """Main Group state file.
+
+    This file contains all the information needed to store the state of a
+    Group object. It includes accessors, setters, and modifiers for all
+    elements of the data structure, as well as the data structure definition.
+
+    """
+    # add new paths to include them in member searches
+    memberpaths = ['abs', 'rel']
+    _fields = ['uuid', 'treanttype', 'abs', 'rel']
+
+    def __init__(self, filename, logger=None, **kwargs):
+        """Initialize Group state file.
+
+        :Arguments:
+           *filename*
+              path to file
+           *logger*
+              logger to send warnings and errors to
+           *categories*
+              user-given dictionary with custom keys and values; used to
+              give distinguishing characteristics to object for search
+           *tags*
+              user-given list with custom elements; used to give distinguishing
+              characteristics to object for search
+        """
+        super(GroupFile, self).__init__(filename, logger=logger, **kwargs)
+
+    def _init_record(self):
+        super(GroupFile, self)._init_record()
+        self._record['members'] = list()
+
+    @FileSerial._write
+    @FileSerial._pull_push
+    def add_member(self, uuid, treanttype, basedir):
+        """Add a member to the Group.
+
+        If the member is already present, its basedir paths will be updated
+        with the given basedir.
+
+        :Arguments:
+            *uuid*
+                the uuid of the new member
+            *treanttype*
+                the treant type of the new member
+            *basedir*
+                basedir of the new member in the filesystem
+
+        """
+        # check if uuid already present
+        uuids = [member[0] for member in self._record['members']]
```

Comment:
That's the other PR #34 (yeah, never doing overlapping PRs ever again; sorry about this). :/

Each member record is now a dict instead of a list, so this problem will go away. Considering merging the changes made for that PR into this branch, but afraid that might just make things more confusing for the moment.